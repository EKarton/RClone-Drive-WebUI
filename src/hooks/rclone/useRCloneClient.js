import RCloneClient from 'services/RCloneClient';
import useRCloneInfo from './useRCloneInfo';

let rCloneInstance = null;

/**
 * Returns the singleton of the RCloneClient
 * If the singleton does not exist yet, it will create one
 *
 * @returns {RCloneClient} the RClone client
 */
export default function useRCloneClient() {
  const { rCloneInfo } = useRCloneInfo();

  if (!rCloneInfo.endpoint || !rCloneInfo.username || !rCloneInfo.password) {
    throw new MissingRCloneInfoError();
  }

  if (rCloneInstance) {
    return rCloneInstance;
  }

  rCloneInstance = new RCloneClient(rCloneInfo, rCloneInfo, rCloneInfo);

  return rCloneInstance;
}

export class MissingRCloneInfoError extends Error {
  constructor() {
    super('Missing RClone info!');
  }
}
