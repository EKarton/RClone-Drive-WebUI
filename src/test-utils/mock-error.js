/**
 * Replaces the error object's stack trace with mock data
 * @param {Error} error the Error object
 * @returns {Error} the same error, but with the stack trace mocked
 */
export function mockErrorStackTrace(error) {
  error.stack = 'Mock content for stack trace';
  return error;
}
