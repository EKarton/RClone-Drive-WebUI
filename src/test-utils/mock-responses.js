export const mockFiles = {
  list: [
    {
      Path: '.rclone',
      Name: '.rclone',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-11-01T21:09:05.237Z',
      IsDir: true,
      ID: '1a4mI4vsGKypqFFPwLnrFid1ZdvlhirDi',
    },
    {
      Path: 'Documents',
      Name: 'Documents',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T01:51:05.982Z',
      IsDir: true,
      ID: '1_za5Kqy9ZjXZ7rsztGpKhtJLTuNCKv8q',
    },
    {
      Path: 'Games',
      Name: 'Games',
      Size: -9,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T05:14:59.237Z',
      IsDir: true,
      ID: '1j6mAjxQoJokiU8yWJYgyWYIiSIgH1G7M',
    },
    {
      Path: 'Literature',
      Name: 'Literature',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-29T07:10:05.059Z',
      IsDir: true,
      ID: '1dj1T5YkjRQZNiQ5efCA8PzideJb88d3K',
    },
    {
      Path: 'Pictures',
      Name: 'Pictures',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:39:11.850Z',
      IsDir: true,
      ID: '1D2PpmoIT23NqyTnolmohZ0UxvOd2D7n4',
    },
    {
      Path: 'Software',
      Name: 'Software',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:45:02.471Z',
      IsDir: true,
      ID: '1zzr0Bwp9tel66rCUKz8mZIXRBR2kuywr',
    },
    {
      Path: 'Videos',
      Name: 'Videos',
      Size: -8,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T04:22:24.997Z',
      IsDir: true,
      ID: '1eJlJ1PT7b9XY3Q5tLeU49BNkug259986',
    },
    {
      Path: 'backup.sh',
      Name: 'backup.sh',
      Size: 1110,
      MimeType: 'application/x-sh',
      ModTime: '2021-11-14T18:58:44.655Z',
      IsDir: false,
    },
  ],
};

export const mockRemotes = {
  remotes: ['googledrive', 'onedrive', 'icloud'],
};

export const mockPictures = {
  list: [
    {
      Path: 'Pictures/2012/20120517_171428.JPG',
      Name: '20120517_171428.JPG',
      Size: 335408,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T21:14:30.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_135206.JPG',
      Name: '20120517_135206.JPG',
      Size: 262719,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T17:52:06.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_204557.JPG',
      Name: '20120516_204557.JPG',
      Size: 245109,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T00:45:58.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_174510.JPG',
      Name: '20120517_174510.JPG',
      Size: 271914,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T21:45:12.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120820_144833.JPG',
      Name: '20120820_144833.JPG',
      Size: 2662880,
      MimeType: 'image/jpeg',
      ModTime: '2012-08-20T18:48:32.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_094409.JPG',
      Name: '20120517_094409.JPG',
      Size: 280596,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T13:44:10.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_181954.JPG',
      Name: '20120516_181954.JPG',
      Size: 416424,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-16T22:19:54.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_181809.JPG',
      Name: '20120516_181809.JPG',
      Size: 211575,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-16T22:18:10.000Z',
      IsDir: false,
    },
  ],
};
