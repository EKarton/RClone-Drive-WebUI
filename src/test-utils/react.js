import { FileViewerProvider } from 'contexts/FileViewerStore';
import { RCloneInfoStateProvider } from 'contexts/RCloneInfoStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState, initialRCloneInfoState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoStateProvider defaultState={initialRCloneInfoState}>
      <FileViewerProvider defaultState={initialFileViewerState}>
        <Router history={history}>{component}</Router>
      </FileViewerProvider>
    </RCloneInfoStateProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
