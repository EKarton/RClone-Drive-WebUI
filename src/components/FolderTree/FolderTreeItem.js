import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import TreeItem from '@mui/lab/TreeItem';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { StatusTypes } from 'utils/constants';

export default function FolderTreeItem({ remote, curPath, label }) {
  const rCloneClient = useRCloneClient();

  const [status, setStatus] = useState(null);
  const [subFolders, setSubFolders] = useState();

  const [isExpanded, setIsExpanded] = useState(false);

  useEffect(() => {
    const fetchFolders = async () => {
      try {
        setStatus(StatusTypes.LOADING);

        const data = await rCloneClient.fetchSubFolders(remote, curPath);

        setSubFolders(data.sort());
        setStatus(StatusTypes.SUCCESS);
      } catch (err) {
        setStatus(StatusTypes.ERROR);
      }
    };

    if (isExpanded) {
      fetchFolders();
    }
  }, [curPath, isExpanded, rCloneClient, remote]);

  const handleTreeItemClicked = () => {
    setIsExpanded(true);
  };

  const renderSubFolders = () => {
    if (!isExpanded || !status) {
      return <div />;
    }

    if (status === StatusTypes.LOADING) {
      return <div>Loading</div>;
    }

    if (status === StatusTypes.ERROR) {
      return <div>Error!</div>;
    }

    if (subFolders.length === 0) {
      return <div>No subfolders</div>;
    }

    return subFolders.map((subFolder) => (
      <FolderTreeItem
        remote={remote}
        curPath={subFolder.Path}
        label={subFolder.Name}
        key={subFolder.Name}
      />
    ));
  };

  return (
    <TreeItem
      nodeId={`${remote}:${curPath}`}
      label={label}
      onClick={handleTreeItemClicked}
      expandIcon={<ChevronRightIcon />}
      collapseIcon={<ExpandMoreIcon />}
    >
      {renderSubFolders()}
    </TreeItem>
  );
}

FolderTreeItem.propTypes = {
  remote: PropTypes.string,
  curPath: PropTypes.string,
  label: PropTypes.string,
};
