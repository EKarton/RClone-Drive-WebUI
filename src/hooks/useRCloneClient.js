import { useMemo } from "react";
import RCloneClient from "utils/RCloneClient";
import useRCloneInfo from "./useRCloneInfo";

export default function useRCloneClient() {
  const { rCloneInfo } = useRCloneInfo();

  if (!rCloneInfo.endpoint || !rCloneInfo.username || !rCloneInfo.password) {
    throw new MissingRCloneInfoError();
  }

  const client = useMemo(() => {
    return new RCloneClient(
      rCloneInfo.endpoint,
      rCloneInfo.username,
      rCloneInfo.password
    );
  }, [rCloneInfo.endpoint, rCloneInfo.username, rCloneInfo.password]);

  return client;
}

export class MissingRCloneInfoError extends Error {
  constructor() {
    super("Missing RClone info!");
  }
}
