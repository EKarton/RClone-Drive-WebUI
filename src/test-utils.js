import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
