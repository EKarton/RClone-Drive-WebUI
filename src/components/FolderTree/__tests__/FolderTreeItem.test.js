import { TreeView } from '@mui/lab';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { customRender, fireEvent, screen } from 'test-utils/react';
import FolderTreeItem from '../FolderTreeItem';

jest.mock('hooks/rclone/useRCloneClient');

// Sub-folders under the '/' directory
const TopLevelFolders = [
  {
    Path: '.rclone',
    Name: '.rclone',
    Size: -2,
    MimeType: 'inode/directory',
    ModTime: '2021-11-01T21:09:05.237Z',
    IsDir: true,
    ID: 'asfsf',
  },
  {
    Path: 'Documents',
    Name: 'Documents',
    Size: -10,
    MimeType: 'inode/directory',
    ModTime: '2021-09-28T01:51:05.982Z',
    IsDir: true,
    ID: 'fsaf',
  },
];

// Sub-folders under the 'Documents' directory
const SecondLevelFolders = [
  {
    Path: 'Work',
    Name: 'Work',
    Size: -2,
    MimeType: 'inode/directory',
    ModTime: '2021-11-01T21:09:05.237Z',
    IsDir: true,
    ID: 'asfsaf',
  },
  {
    Path: 'Personal',
    Name: 'Personal',
    Size: -10,
    MimeType: 'inode/directory',
    ModTime: '2021-09-28T01:51:05.982Z',
    IsDir: true,
    ID: 'sakfmkasfm',
  },
];

describe('FolderTreeItem', () => {
  const fetchSubFoldersFn = jest.fn();

  beforeEach(() => {
    useRCloneClient.mockReturnValue({
      fetchSubFolders: fetchSubFoldersFn,
    });
  });

  it('should make api calls match snapshots when user clicks on subfolders', async () => {
    fetchSubFoldersFn.mockImplementation((remote, curPath) => {
      if (curPath === '') {
        return Promise.resolve(TopLevelFolders);
      }

      return Promise.resolve(SecondLevelFolders);
    });

    const { baseElement } = customRender(
      <TreeView>
        <FolderTreeItem remote="gdrive" curPath="" label="/" />
      </TreeView>
    );

    await screen.findByText('/');
    expect(baseElement).toMatchSnapshot();

    fireEvent.click(screen.getByText('/'));

    await screen.findByText('Documents');
    expect(baseElement).toMatchSnapshot();

    fireEvent.click(screen.getByText('Documents'));

    await screen.findByText('Personal');
    expect(baseElement).toMatchSnapshot();
  });

  it('should render loading and then display data when api is loading and has succeeded', async () => {
    // Mock the timer
    jest.useFakeTimers();

    // Mock api call that takes 10000ms
    fetchSubFoldersFn.mockImplementation(() => {
      return new Promise((resolve) => {
        setTimeout(() => {
          resolve(TopLevelFolders);
        }, 10000);
      });
    });

    customRender(
      <TreeView>
        <FolderTreeItem remote="gdrive" curPath="" label="/" />
      </TreeView>
    );

    await screen.findByText('/');
    fireEvent.click(screen.getByText('/'));

    // Resolve api call
    jest.runAllTimers();

    await screen.findByText('Documents');
  });

  it('should render error when api call failed', async () => {
    fetchSubFoldersFn.mockRejectedValue(new Error('Random error'));

    customRender(
      <TreeView>
        <FolderTreeItem remote="gdrive" curPath="" label="/" />
      </TreeView>
    );

    await screen.findByText('/');
    fireEvent.click(screen.getByText('/'));

    await screen.findByText('Error!');
  });

  it('should render correctly when there are no more sub-folders', async () => {
    fetchSubFoldersFn.mockResolvedValue([]);

    customRender(
      <TreeView>
        <FolderTreeItem remote="gdrive" curPath="" label="/" />
      </TreeView>
    );

    await screen.findByText('/');
    fireEvent.click(screen.getByText('/'));

    await screen.findByText('No subfolders');
  });
});
