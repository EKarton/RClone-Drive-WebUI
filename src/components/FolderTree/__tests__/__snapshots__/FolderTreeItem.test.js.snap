// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderTreeItem should make api calls match snapshots when user clicks on subfolders 1`] = `
<body>
  <div>
    <ul
      aria-multiselectable="false"
      class="MuiSimpleTreeView-root css-rejl51-MuiSimpleTreeView-root"
      id=":r0:"
      role="tree"
      style="--TreeView-itemChildrenIndentation: 12px;"
    >
      <li
        aria-expanded="false"
        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
        collapseicon="[object Object]"
        expandicon="[object Object]"
        id=":r0:-gdrive:"
        role="treeitem"
        tabindex="0"
      >
        <div
          class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content"
        >
          <div
            class="MuiTreeItem-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
              data-testid="TreeViewExpandIconIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </div>
          <div
            class="MuiTreeItem-label"
          >
            /
          </div>
        </div>
      </li>
    </ul>
  </div>
</body>
`;

exports[`FolderTreeItem should make api calls match snapshots when user clicks on subfolders 2`] = `
<body>
  <div>
    <ul
      aria-multiselectable="false"
      class="MuiSimpleTreeView-root css-rejl51-MuiSimpleTreeView-root"
      id=":r0:"
      role="tree"
      style="--TreeView-itemChildrenIndentation: 12px;"
    >
      <li
        aria-expanded="true"
        aria-selected="true"
        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
        collapseicon="[object Object]"
        expandicon="[object Object]"
        id=":r0:-gdrive:"
        role="treeitem"
        tabindex="0"
      >
        <div
          class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content Mui-expanded Mui-selected Mui-focused"
        >
          <div
            class="MuiTreeItem-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
              data-testid="TreeViewCollapseIconIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </div>
          <div
            class="MuiTreeItem-label"
          >
            /
          </div>
        </div>
        <ul
          class="MuiCollapse-root MuiCollapse-vertical MuiTreeItem-groupTransition css-ugkyyp-MuiCollapse-root-MuiTreeItem-groupTransition"
          role="group"
          style="min-height: 0px; height: 0px; transition-duration: 300ms;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
            >
              <li
                aria-expanded="false"
                class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
                collapseicon="[object Object]"
                expandicon="[object Object]"
                id=":r0:-gdrive:.rclone"
                role="treeitem"
                tabindex="-1"
              >
                <div
                  class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                      data-testid="TreeViewExpandIconIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiTreeItem-label"
                  >
                    .rclone
                  </div>
                </div>
              </li>
              <li
                aria-expanded="false"
                class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
                collapseicon="[object Object]"
                expandicon="[object Object]"
                id=":r0:-gdrive:Documents"
                role="treeitem"
                tabindex="-1"
              >
                <div
                  class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                      data-testid="TreeViewExpandIconIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiTreeItem-label"
                  >
                    Documents
                  </div>
                </div>
              </li>
            </div>
          </div>
        </ul>
      </li>
    </ul>
  </div>
</body>
`;

exports[`FolderTreeItem should make api calls match snapshots when user clicks on subfolders 3`] = `
<body>
  <div>
    <ul
      aria-multiselectable="false"
      class="MuiSimpleTreeView-root css-rejl51-MuiSimpleTreeView-root"
      id=":r0:"
      role="tree"
      style="--TreeView-itemChildrenIndentation: 12px;"
    >
      <li
        aria-expanded="true"
        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
        collapseicon="[object Object]"
        expandicon="[object Object]"
        id=":r0:-gdrive:"
        role="treeitem"
        tabindex="-1"
      >
        <div
          class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content Mui-expanded"
        >
          <div
            class="MuiTreeItem-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
              data-testid="TreeViewCollapseIconIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </div>
          <div
            class="MuiTreeItem-label"
          >
            /
          </div>
        </div>
        <ul
          class="MuiCollapse-root MuiCollapse-vertical MuiTreeItem-groupTransition css-ugkyyp-MuiCollapse-root-MuiTreeItem-groupTransition"
          role="group"
          style="min-height: 0px; height: 0px; transition-duration: 300ms;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
            >
              <li
                aria-expanded="false"
                class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
                collapseicon="[object Object]"
                expandicon="[object Object]"
                id=":r0:-gdrive:.rclone"
                role="treeitem"
                tabindex="-1"
              >
                <div
                  class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                      data-testid="TreeViewExpandIconIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiTreeItem-label"
                  >
                    .rclone
                  </div>
                </div>
              </li>
              <li
                aria-expanded="true"
                aria-selected="true"
                class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
                collapseicon="[object Object]"
                expandicon="[object Object]"
                id=":r0:-gdrive:Documents"
                role="treeitem"
                tabindex="0"
              >
                <div
                  class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content Mui-expanded Mui-selected Mui-focused"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                      data-testid="TreeViewCollapseIconIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiTreeItem-label"
                  >
                    Documents
                  </div>
                </div>
                <ul
                  class="MuiCollapse-root MuiCollapse-vertical MuiTreeItem-groupTransition css-ugkyyp-MuiCollapse-root-MuiTreeItem-groupTransition"
                  role="group"
                  style="min-height: 0px; height: 0px; transition-duration: 300ms;"
                >
                  <div
                    class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                  >
                    <div
                      class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                    >
                      <li
                        aria-expanded="false"
                        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
                        collapseicon="[object Object]"
                        expandicon="[object Object]"
                        id=":r0:-gdrive:Work"
                        role="treeitem"
                        tabindex="-1"
                      >
                        <div
                          class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content"
                        >
                          <div
                            class="MuiTreeItem-iconContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                              data-testid="TreeViewExpandIconIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                              />
                            </svg>
                          </div>
                          <div
                            class="MuiTreeItem-label"
                          >
                            Work
                          </div>
                        </div>
                      </li>
                      <li
                        aria-expanded="false"
                        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
                        collapseicon="[object Object]"
                        expandicon="[object Object]"
                        id=":r0:-gdrive:Personal"
                        role="treeitem"
                        tabindex="-1"
                      >
                        <div
                          class="css-53rj3a-MuiTreeItem-content MuiTreeItem-content"
                        >
                          <div
                            class="MuiTreeItem-iconContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                              data-testid="TreeViewExpandIconIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                              />
                            </svg>
                          </div>
                          <div
                            class="MuiTreeItem-label"
                          >
                            Personal
                          </div>
                        </div>
                      </li>
                    </div>
                  </div>
                </ul>
              </li>
            </div>
          </div>
        </ul>
      </li>
    </ul>
  </div>
</body>
`;
